src/Agent.Plugins/PipelineArtifact/FilePathUNCServer.cs

```diff
@@ -0,0 +1,95 @@
+                { FileShareArtifactUploadEventProperties.ArtifactLocation, fileSharePath }
+            };
+
+            var artifact = await buildServer.AssociateArtifactAsync(projectId, buildId, artifactName, ArtifactResourceTypes.FilePath, fileSharePath, propertiesDictionary, token);
+            var parallel = context.GetInput(FileShareArtifactUploadEventProperties.Parallel, required: false);
+            var parallelCount = parallel == "true" ? GetParallelCount(context, context.GetInput(FileShareArtifactUploadEventProperties.ParallelCount, required: false)) : 1;
+
+            // To copy all the files in one directory to another directory.
+            // Get the files in the source folder. (To recursively iterate through
+            // all subfolders under the current directory, see
+            // "How to: Iterate Through a Directory Tree.")
+            // Note: Check for target path was performed previously
+            // in this code example.
+            if (System.IO.Directory.Exists(fileSharePath))
+            {
+                FileShareProvider provider = new FileShareProvider(context, new CallbackAppTraceSource(str => context.Output(str), System.Diagnostics.SourceLevels.Information));
+                await provider.PublishArtifactAsync(targetPath, artifactPath, parallelCount, token);
+                context.Output(StringUtil.Loc("CopyFileComplete", artifactPath));
+            }
+        }
+
+        internal static class FileShareArtifactUploadEventProperties
+        {
+            public const string ArtifactName = "artifactname";
+            public const string ArtifactLocation = "artifactlocation";
+            public const string ArtifactType = "artifacttype";
+            public const string ParallelCount = "parallelCount";
+            public const string Parallel = "parallel";
+        }
+
+        // Enter the degree of parallelism, or number of threads used, to perform the copy. The value must be at least 1 and not greater than 128.
+        // This is the same logic as the build artifact tasks https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts
+        private int GetParallelCount(AgentTaskPluginExecutionContext context, string parallelCount)
+        {
+            var result = 8;
+            if(int.TryParse(parallelCount, out result))
+            {
+                if(result < 1) 
+                {
+                    context.Output(StringUtil.Loc("UnexpectedParallelCount"));
+                    result = 1;
+                }
+                else if(result > 128)
+                {
+                    context.Output(StringUtil.Loc("UnexpectedParallelCount"));
+                    result = 128;
+                }
+            }
+            else 
+            {
+                throw new ArgumentException(StringUtil.Loc("ParallelCountNotANumber"));
```

Comment:
does this need to happen?

Reply:
why not just fall back to default and move on?

---
In reply to: [316905244](https://github.com/microsoft/azure-pipelines-agent/pull/2385#discussion_r316905244) [](ancestors = 316905244)

Reply:
This behavior is as the same as build artifact task. (Link to the code is provided in the comment in the code)